types.go

```diff
@@ -83,3 +64,91 @@ func (m PubKey) AsABCI() abci.PubKey {
 		Type: m.Type,
 	}
 }
+
+func (m ValidatorUpdates) Validate() error {
+	var err error
+	for _, v := range m.ValidatorUpdates {
+		err = errors.Append(err, v.Validate())
+	}
+	return err
+}
+
+// Get gets a ValidatorUpdate by a public key if it exists
+func (m ValidatorUpdates) Get(key PubKey) (ValidatorUpdate, int, bool) {
+	for k, v := range m.ValidatorUpdates {
+		if v.PubKey.Type == key.Type && bytes.Equal(v.PubKey.Data, key.Data) {
+			return v, k, true
+		}
+	}
+
+	return ValidatorUpdate{}, -1, false
+}
+
+// Deduplicate makes sure we only use the last validator update to any given validator.
+// For bookkeeping we have an option to drop validators with zero power, because those
+// are being removed by tendermint once propagated.
+func (m ValidatorUpdates) Deduplicate(dropZeroPower bool) ValidatorUpdates {
+	if len(m.ValidatorUpdates) == 0 {
+		return m
+	}
+
+	duplicates := make(map[string]int)
```

Comment:
note: it took me a bit to understand this deduplication proper. `duplicate` is basically an index that points to a position (instead of `key`) within the result **Set**: `deduplicatedValidators` correct? Could you please add some comments/ renaming as this may be complicated for others as well.